### ISSUE TITLE
grapher usage question

### ISSUE THREAD

**Issue by ujwal-setlur** [cult-of-coders/apollo#40]
Hi,

I have a question on how to setup grapher and links. Let's say I have two collections: Users and Groups. Groups will have a a field called memberIds that is an array of Users. 

What I want to do is get a list of all the groups I am a member of, and get the **firstName** of all the other users of the groups as well. I set up the schema/links in this way:

```
type Group @mongo(name: "groups")
{
  _id: ID!
  name: String
  members: [User] @link(field: "memberIds")
  admins: [User] @link(field: "adminIds")
}

# our user type
type User @mongo(name: "users")
{
  _id: ID!
  emails: [Email]
  firstName: String
  lastName: String
  fullName: String
  roles: [String]
  groups: [Group] @link(to: "members")
  admins: [Group] @link(to: "admins")
}
```

Here is my GraphQL query:

```
query groups {
  groups{
    _id
    name
    members {
      _id
      firstName
      roles
    }
  }
}
```

The issue I am facing is that the first name of user is not being retrieved:

```
{
  "data": {
    "groups": [
      {
        "_id": "YqzhXSWYp6BXGBRkz",
        "name": "Test Group",
        "members": [
          {
            "_id": "PHpfwhHDR25koG5Bd",
            "firstName": null,
            "roles": [
              "seamlz.factory"
            ]
          }
        ]
      }
    ]
  },
  "errors": [
    {
      "message": "Cannot read property 'firstName' of undefined",
      "locations": [
        {
          "line": 7,
          "column": 7
        }
      ],
      "path": [
        "groups",
        0,
        "members",
        0,
        "firstName"
      ]
    }
  ]
}
```

The first name is actually buried in the profile object in the user collection:

```js
{
	"_id" : "PHpfwhHDR25koG5Bd",
	"createdAt" : ISODate("2018-07-31T07:34:47.128Z"),
	"services" : {
		"phone" : {

		},
		"resume" : {
			"loginTokens" : [
				{
					"when" : ISODate("2018-07-31T07:34:55.384Z"),
					"hashedToken" : "65jCfZi4ItIU3OwQyhUEZg158QUXp21vV6thfYV/31c="
				}
			]
		}
	},
	"phone" : {
		"number" : "+19253174023",
		"verified" : true
	},
	"profile" : {
		"name" : "Vineet Chikarmane",
		"firstName" : "Vineet",
		"lastName" : "Chikarmane",
		"company" : "Pret Interpret"
	},
	"roles" : [
		"seamlz.factory"
	]
}
```

How do I get grapher to dig into the profile section to retrieve the firstName?

Thanks!

**Comment by theodorDiaconu** [comment]
@ujwal-setlur yes I understand, the problem now lies in tying GraphQL with Grapher. Because you have `firstName` stored in User, and most likely I'm guessing you have a resolver that reads from profile that firstName, yet you put it at top level in User.

You have the following options:

1. (Recommended) Create a Reducer for Profile Name
```js
Users.addReducers({
    'firstName': {
         body: { 'profile.firstName': 1 }
    },
    reduce: obj => obj.profile.firstName
})
```

Now you can safely use it.

2. Inject profile in your query via `embody`: https://github.com/cult-of-coders/grapher/blob/master/docs/graphql.md#creating-your-queries

This is what we do with @map directive from https://github.com/cult-of-coders/grapher-schema-directives, but I created an issue: https://github.com/cult-of-coders/grapher-schema-directives/issues/8 and will solve it.

**Comment by ujwal-setlur** [comment]
@theodorDiaconu, thanks I will try both options.

**Comment by theodorDiaconu** [comment]
@ujwal-setlur solved in grapher-schema-directives@0.1.6

```js
firstName: String @map("profile.firstName")
```

Elegant and automatic. I will update deps here as well in my next iteration. I have some nice things to add.

**Comment by ujwal-setlur** [comment]
@theodorDiaconu, fantastic! I tried both your options, and both worked! I had actually tried the ```embody``` options last night, but I was not injecting profile in the right place. Thank you so much. I keep being amazed how powerful this stack is. Kudos!

**Comment by ujwal-setlur** [comment]
We can close this I think.

**Comment by ujwal-setlur** [comment]
@theodorDiaconu unfortunately, the new map directive for nested fields in grapher-schema-directives 0.1.6 does not work:

```
firstName: String @map("profile.firstName")
```

I get a build error. I get:

```
imports/api/users/user.graphql:18:26: Syntax Error: Expected Name, found String "profile.firstName"
```

Then I tried

```
firstName: String @map(field: "profile.firstName")
```

This time, the app builds, but the query fails:

```
query users {
  users{
    _id
    name
    company
    roles
    groups {
      _id
      name
      description
    }
  }
}
```

I get:

```
{
  "data": {
    "users": null
  },
  "errors": [
    {
      "message": "object is not defined",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "users"
      ]
    }
  ]
}
```

**Comment by ujwal-setlur** [comment]
@theodorDiaconu I found a bug in grapher-schema-directives 0.1.6. I have opened a pull request:

https://github.com/cult-of-coders/grapher-schema-directives/pull/10

**Comment by theodorDiaconu** [comment]
It was fixed and released 0.1.7. Sorry about that.
